libgit/util.go

```diff
@@ -0,0 +1,75 @@
+// Copyright 2018 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libgit
+
+import (
+	"context"
+	"os"
+	"time"
+
+	"github.com/keybase/client/go/logger"
+	"github.com/keybase/kbfs/libfs"
+	"github.com/keybase/kbfs/libkbfs"
+)
+
+// commonTime computes the current time according to our estimate of
+// the mdserver's time.  It's a very crude way of normalizing the
+// local clock.
+func commonTime(
+	ctx context.Context, mdserver libkbfs.MDServer, clock libkbfs.Clock,
+	log logger.Logger) time.Time {
+	offset, haveOffset := mdserver.OffsetFromServerTime()
+	if !haveOffset {
+		log.CDebugf(ctx, "No offset, cannot use common time; "+
+			"falling back to local time")
+		return clock.Now()
+	}
+	return clock.Now().Add(-offset)
+}
+
+// canDoWork creates a file that marks the start of some long-term
+// work by this node.  It should be called while a lock is taken on
+// the server.  It returns true if the caller should start doing the
+// work.
+func canDoWork(
+	ctx context.Context, mdserver libkbfs.MDServer, clock libkbfs.Clock,
+	fs *libfs.FS, workingFileName string, workLimit time.Duration,
+	log logger.Logger) (bool, error) {
+	fi, err := fs.Stat(workingFileName)
+	currCommonTime := commonTime(ctx, mdserver, clock, log)
+	if err != nil && !os.IsNotExist(err) {
+		return false, err
+	} else if os.IsNotExist(err) {
+		log.CDebugf(ctx, "Creating new working file %s", workingFileName)
+		f, err := fs.Create(workingFileName)
+		if err != nil {
+			return false, err
+		}
+		err = f.Close()
+		if err != nil {
+			return false, err
+		}
+	} else { // err == nil
+		modCommonTime := fi.ModTime()
+		if modCommonTime.Add(workTimeLimit).After(currCommonTime) {
+			log.CDebugf(ctx, "Other worker is still working; "+
+				"modCommonTime=%s, currCommonTime=%s, workTimeLimit=%s",
+				modCommonTime, currCommonTime, workTimeLimit)
+			// The other GC is still running within the time
+			// limit.
+			return false, nil
+		}
+		log.CDebugf(ctx, "Other GC expired; "+
+			"modCommonTime=%s, currCommonTime=%s, workTimeLimit=%s",
+			modCommonTime, currCommonTime, workTimeLimit)
+	}
```

Comment:
I know this is from the other file, but it'd be nice if this was in a switch/case block.